### PR TITLE
[Fix #952] Exclude out-of-project VCS submodules

### LINKED ISSUE #952: VCS submodules brought in even thought not descendent of project root
Given the following tree structure:

```
./.git # git repo top-level
./front-end/
./front-end/.projectile
./front-end/src/...
./server/
./server/vendor
./server/vendor/a-git-submodule/
./server/vendor/a-git-submodule/.git
```

If the current top-level folder is `./front-end/`, and you invoke `projectile-find-file`, then `../server/vendor/a-git-submodule/` will leak into your choices. Presently, all submodules for the current repository are traversed, recursively so, regardless of whether or not those submodules reside within the current project root.

### PR DESCRIPTION
**Pull request by timcharper**
Given the following tree structure:

```
./.git # git repo top-level
./front-end/
./front-end/.projectile
./front-end/src/...
./server/
./server/vendor
./server/vendor/a-git-submodule/
./server/vendor/a-git-submodule/.git
```

If the current top-level folder is `./front-end/`, and you invoke `projectile-find-file`, then `../server/vendor/a-git-submodule/` will leak into your choices. Presently, all submodules for the current repository are traversed, recursively so, regardless of whether or not those submodules reside within the current project root.

The attached patch resolves that issue.

Thank you for projectile!

### CODE REVIEW

**Comment by timcharper** [comment]
@bbatsov can you please review this pull request? I've been using the patch for several months now and it does indeed fix a verifiable bug. Would you like a repro case?

**Comment by timcharper** [comment]
I've rebased by pull request to the latest master.

**Comment by timcharper** [comment]
Also, I noticed that `cl` functions are not allowed. Switched dependency over to dash and everything passes now.

**Comment by timcharper** [comment]
I added a test.

**Comment by bbatsov** [comment]
Sorry for the late response. I'll try to review this in the next day or two. I've been pretty short on time lately.

**Comment by timcharper** [comment]
Totally understand! Hope that adding a test helps. I can provide a repository to clone to reproduce the issue, also, if that'd help.

**Comment by timcharper** [comment]
Anything I can do more to help get this patch made? Want me to hop on a plane and come wash your dishes while you merge my patch? lol

**Comment by bbatsov** [comment]
I've got a dishwasher. ;-) I've been trying to find a bit of time for projectile lately, but to no avail. I'll try to review this later today or tomorrow.

File: projectile.el [hunk 1]
```diff
@@ -867,16 +867,31 @@ Files are returned as relative paths to the project root."
      ((eq vcs 'git) projectile-git-submodule-command)
      (t ""))))
 
-(defun projectile-get-all-sub-projects (project)
-  "Get all sub-projects for a given projects.
-PROJECT is base directory to start search recursively."
+(defun projectile-get-immediate-sub-projects (project)
+  "Get immediate sub-projects for a given project. Does not recurse.
```
Comment (bbatsov) [inline projectile.el]: `M-x checkdoc`

* The first line should be just a single sentence
* the final sentence doesn't end with a `.`.
* PROJECT should appear somewhere in the docstring (params should be mentioned in all caps)

File: projectile.el [hunk 2]
```diff
@@ -867,16 +867,31 @@ Files are returned as relative paths to the project root."
...
+It is assumed that `project` will end with an appropriate path delimiter, such as '/' or a '\'"
   (let* ((default-directory project)
          ;; search for sub-projects under current project `project'
          (submodules (mapcar
                       (lambda (s)
-                        (file-name-as-directory (expand-file-name s default-directory)))
-                      (projectile-files-via-ext-command (projectile-get-sub-projects-command)))))
+                        (file-name-as-directory (expand-file-name s project)))
```
Comment (bbatsov) [inline projectile.el]: default-directory and project is the same thing here - notice the first binding

File: projectile.el [hunk 3]
```diff
@@ -867,16 +867,31 @@ Files are returned as relative paths to the project root."
...
+                      (projectile-files-via-ext-command (projectile-get-sub-projects-command))))
+         (project-child-folder-regex
+          (concat "^"
```
Comment (bbatsov) [inline projectile.el]: Start of string is `\\`` in Emacs Lisp.

File: projectile.el [hunk 4]
```diff
@@ -867,16 +867,31 @@ Files are returned as relative paths to the project root."
      ((eq vcs 'git) projectile-git-submodule-command)
      (t ""))))
 
-(defun projectile-get-all-sub-projects (project)
-  "Get all sub-projects for a given projects.
-PROJECT is base directory to start search recursively."
+(defun projectile-get-immediate-sub-projects (project)
+  "Get immediate sub-projects for a given project. Does not recurse.
+It is assumed that `project` will end with an appropriate path delimiter, such as '/' or a '\'"
```
Comment (bbatsov) [inline projectile.el]: Guess you should validate the project name, instead of just assuming it ends the way you expect it to end. :-)

File: projectile.el [hunk 5]
```diff
@@ -867,16 +867,31 @@ Files are returned as relative paths to the project root."
...
                       (lambda (s)
-                        (file-name-as-directory (expand-file-name s default-directory)))
-                      (projectile-files-via-ext-command (projectile-get-sub-projects-command)))))
+                        (file-name-as-directory (expand-file-name s project)))
+                      (projectile-files-via-ext-command (projectile-get-sub-projects-command))))
+         (project-child-folder-regex
+          (concat "^"
+                  (regexp-quote default-directory))))
+
+    ;; If project root is inside of an SCM folder, but not actually an
```
Comment (bbatsov) [inline projectile.el]: SCM -> VCS (for consistency with the rest of the code)

File: projectile.el [hunk 6]
```diff
@@ -867,16 +867,31 @@ Files are returned as relative paths to the project root."
...
+    ;; SCM root itself, external paths will leak in. Let's remove them
+    (-filter (lambda (path)
+               (string-match-p project-child-folder-regex
+                               path))
+             submodules)))
+
+(defun projectile-get-all-sub-projects (project)
```
Comment (bbatsov) [inline projectile.el]: This should come before the other definition to avoid a compiler warning about using something that's not yet defined.

File: projectile.el [hunk 7]
```diff
@@ -867,16 +867,31 @@ Files are returned as relative paths to the project root."
...
+  "Get all sub-projects for a given project.
+PROJECT is base directory to start search recursively."
```
Comment (bbatsov) [inline projectile.el]: Kill this blank line.

File: test/projectile-test.el [hunk 8]
```diff
@@ -636,6 +636,28 @@
...
 
+(ert-deftest projectile-test-exclude-out-of-project-submodules ()
+  (projectile-test-with-files
+      ("project/.git/"
+       "project/.gitmodules"
+       "project/web-ui/"
+       "project/web-ui/.projectile"
+       "project/web-ui/vendor/client-submodule/"
+       "project/server/vendor/server-submodule/")
+    (let ((project (file-truename (expand-file-name "project/web-ui")))
```
Comment (bbatsov) [inline test/projectile-test.el]: Is `file-truename` needed here?

File: test/projectile-test.el [hunk 9]
```diff
@@ -636,6 +636,28 @@
...
+          (projectile-indexing-method 'native))
```
Comment (bbatsov) [inline test/projectile-test.el]: Native? Are you sure about this?

Comment (bbatsov) [inline test/projectile-test.el]: I think I understand your reasoning but the test is a bit hard to follow. Adding more clarity to it would be great.

File: projectile.el (hunk 4 above)
Comment (timcharper) [inline projectile.el]: That seems excessive. (IE we wouldn't go and validate that a number was a number everywhere a number was expected).

It's just describing the expected schema.

File: projectile.el (hunk 2 above)
Comment (timcharper) [inline projectile.el]: Okay; sure, there's no reason to change from `default-directory` to `project`. No problem.

File: projectile.el (hunk 1 above)
Comment (timcharper) [inline projectile.el]: done

File: projectile.el (hunk 3 above)
Comment (timcharper) [inline projectile.el]: Both work but that's slightly better. Done.

File: projectile.el (hunk 5 above)
Comment (timcharper) [inline projectile.el]: done

File: projectile.el (hunk 6 above)
Comment (timcharper) [inline projectile.el]: transposed

File: test/projectile-test.el (hunk 9 above)
Comment (timcharper) [inline test/projectile-test.el]: Shoot; I was wrong. When I wrote the test I was thinking `native` meant "use OS native", not "use Emacs native implementation".

I definitely did not mean native.

File: test/projectile-test.el (hunk 8 above)
Comment (timcharper) [inline test/projectile-test.el]: Possibly? You could be running the tests from within path that is a symlink and I didn't know if things would break.

File: test/projectile-test.el (hunk 9 above)
Comment (timcharper) [inline test/projectile-test.el]: Man, I checked to make sure my patch failed without my test. I'm not sure how it worked with it set to `native`.

Comment (timcharper) [inline test/projectile-test.el]: Oh, it's because I'm calling `projectile-get-all-sub-projects` directly, so it doesn't matter. I'll just not set it since it's irrelevant.

File: projectile.el (hunk 7 above)
Comment (timcharper) [inline projectile.el]: done

**Comment by timcharper** [comment]
Sigh... the test ran locally... I'll get it working.

File: projectile.el [hunk 10]
```diff
@@ -912,15 +912,9 @@ Files are returned as relative paths to the project root."
      (t ""))))
 
 (defun projectile-get-all-sub-projects (project)
-  "Get all sub-projects for a given projects.
+  "Get all sub-projects for a given project.
```
Comment (bbatsov) [inline projectile.el]: project -> PROJECT

**Comment by timcharper** [comment]
@bbatsov everything is passing again and I've incorporated your changes. Can you give it another look?

File: projectile.el [hunk 11]
```diff
@@ -930,6 +925,33 @@ PROJECT is base directory to start search recursively."
...
 
+(defun projectile-get-immediate-sub-projects (path)
+  "Get immediate sub-projects for a given project without recursing.
+
+PATH is the vcs root or project root from which to start
+searching, and should end with an appropriate path delimiter, such as
+'/' or a '\\'.
+
+If the vcs get-sub-projects query returns results outside of path,
+they are excluded from the results of this function"
```
Comment (bbatsov) [inline projectile.el]: You forgot the final `.` here.

**Comment by bbatsov** [comment]
The changes look good. Just add a changelog entry and we're good to go.

**Comment by timcharper** [comment]
Done and done.

**Comment by bbatsov** [comment]
Hmm, there's some merge conflict now. Guess you have to rebase on top of the current `master.`

**Comment by timcharper** [comment]
It was a conflict in CHANGES.md. resolved

**Comment by bbatsov** [comment]
:+1:

**Comment by jonEbird** [comment]
@timcharper, I pulled the latest version with your commit merged but am still getting submodule results. I assumed a prefix'ed call to `projectile-find-file` would be all that was required?

**Comment by timcharper** [comment]
That's weird. If you can reproduce it in a bare bones git repo and share it so I can try and reproduce your issue I might be able to help.
